fake.go

```diff
@@ -0,0 +1,161 @@
+	// not required, particularly if the function is not a method.
+	Receiver interface{}
+
+	// Funcname is the name of the function that was called.
+	FuncName string
+
+	// Args is the set of arguments to the function.
+	Args FakeCallArgs
+}
+
+// Fake is used in testing to stand in for some other value, to record
+// all calls to faked methods/functions, and to allow users to set the
+// values that are returned from those calls. Fake is intended to be
+// embedded in another struct that will define the methods to track:
+//
+//    type fakeConn struct {
+//        *testing.Fake
+//        Response []byte
+//    }
+//
+//    func newFakeConn() *fakeConn {
+//        return &fakeConn{
+//            Fake: &testing.Fake{},
+//        }
+//    }
+//
+//    func (fc fakeConn) Send(request string) []byte {
+//        fc.AddCall("Send", FakeCallArgs{
+//            "request": request,
+//        })
+//        return fc.Response, fc.Err()
+//    }
+//
+// As demonstrated in the example, embed a pointer to testing.Fake. This
+// allows a single testing.Fake to be shared between multiple fakes.
+//
+// Error return values are set through Fake.Errors. Set it to the errors
+// you want returned. The errors will be matched up to the calls made on
+// Fake methods, in order. This is facilitated by the Err method, as
+// seen in the above example. In some cases the first method call is not
+// the one you want to fail. If not then put a nil before the error in
+// Fake.Errors. Fake.SetErrors is a helper for setting up failure cases.
+//
+// To validate calls made to the fake in a test, call the CheckCalls
+// method:
+//
+//    s.fake.CheckCalls(c, []FakeCall{{
+//        FuncName: "Send",
+//        Args: FakeCallArgs{
+//            "request": expected,
+//        },
+//    }
+//
+// Not only is Fake useful for building a interface implementation to
+// use in testing (e.g. a network API client), it is also useful in
+// patching situations:
+//
+//    s.PatchValue(&somefunc, s.fake.Send)
+//
+// This allows for easily monitoring the args passed to the patched
+// func, as well as controlling the return value from the func in a
+// clean manner (by simply setting the correct fake field).
+type Fake struct {
+	// Calls is the list of calls that have been registered on the fake
+	// (i.e. made on the fake's methods), in the order that they were
+	// made.
+	Calls []FakeCall
+
+	// Errors holds the list of error return values to use for
+	// successive calls to methods that return an error. Each call
+	// pops the next error off the list. An empty list (the default)
+	// implies a nil error. nil may be precede actual errors in the
+	// list, which means that the first calls will succeed, followed
+	// by the failure. All this is facilitated through the Err method.
+	Errors []error
+
+	// Error is the default error (when Errors is empty). The typical
+	// Fake usage will leave this nil (i.e. no error).
+	Error error
```

Comment:
Maybe call this DefaultError?  Error vs. Errors is not very obvious unless you're reading the documentation.